text/0024-workflows.md

```diff
@@ -0,0 +1,299 @@
+the `Workflow` object either explicitly by hard-coding the workflow definition
+in code or by loading the workflow definition from state (a future smart
+contract may be provided to make storing workflows a generalized Grid
+capability, but it is out-of-scope for this RFC).
+
+The `Workflow` contains a list of `SubWorkflow` objects representing different
+smaller workflows that make up the overall workflow. Each `SubWorkflow` contains
+a `WorkflowState` that describes the possible state transitions the
+`SubWorkflow` can make, the constraints that need to be met to make said
+transitions, and a list of `PermissionAlias` objects that represent the
+permissions that are required by the acting entity to initiate a transition.
+Finally, a `PermissionAlias` object is an object that represents a list of
+permissions and the transitions those permissions are allowed to make. These
+permissions are initially defined in the Pike Smart Contract. For example, a
+permission alias called `po.seller`, meant to represent the seller listed on a
+purchase order, is an alias for the permissions to create, and update a purchase
+order, and can move the purchase order into the confirmed state.
+
+```
+let alias = PermissionAlias {
+   name: “po.seller”.to_string(),
+   permissions: vec![“update_po”.to_string(), “update_version”.to_string()],
+   transitions: vec![“confirmed”.to_string()]
+}
+```
+
+Below is an example of how to instantiate a simple purchase order workflow.
+
+```
+fn get_po_workflow() -> Workflow {
+    let mut permission = PermissionAlias::new("po.seller");
+    permission.add_permission("po.create");
+    permission.add_permission("po.update");
+    permission.add_transition("confirm");
+
+    let state = WorkflowStateBuilder::new("issued")
+        .add_constraint("active=None")
+        .add_transition("issued")
+        .add_transition("confirm")
+        .add_permission_alias(permission)
+        .build();
+
+    let subworkflow = SubWorkflowBuilder::new("po")
+        .add_state(state)
+        .add_starting_state("issued")
+        .add_starting_state("proposed")
+        .build();
+
+    let workflow = Workflow::new(vec![subworkflow]);
+}
+```
+
+Below is an example outlining this process for a hypothetical smart contract designed to process purchase orders.
+
+```
+fn apply(
+    &self,
+    request: &TpProcessRequest,
+    context: &mut dyn TransactionContext,
+) -> Result<(), ApplyError> {
+   let current_state = context.get_state_entry(PO_STATE_ADDRESS);
+   let desired_state = request.get_state();
+   let workflow = get_po_workflow();
+   let subworkflow = workflow.subworkflow("po");
+   let state = subworkflow.state(current_state);
+   let pike_perms = context.get_state_entry(PIKE_ADDRESS);
+   let permissions = state.expand_permissions(pike_perms);
+
+   let can_transition = state.can_transition(
+       desired_state, pike_permissions);
+   if (can_transition) {
```

Comment:
s/if (can_transition)/if (!can_transition)/ ?

Reply:
Updated to use the exclamation